lib/os_io.c

```diff
@@ -37,6 +37,197 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_select_or_poll
+
+___HIDDEN int ___fdset_init
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+
+  int size = ___VMSTATE_FROM_PSTATE(ps)->os.fdset.size;
+  int bytes = ___CEILING_DIV (size, 8);
+
+  readfds  = ___ALLOC_MEM (bytes);
+  if (readfds == NULL)
+    goto error;
+
+  writefds = ___ALLOC_MEM (bytes);
+  if (writefds == NULL)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (bytes);
+  if (exceptfds == NULL)
+    goto error;
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = size;
+
+  return 1;
+
+ error:
+  if (readfds != NULL)
+    ___FREE_MEM (readfds);
+  if (writefds != NULL)
+    ___FREE_MEM (writefds);
+  if (exceptfds != NULL)
+    ___FREE_MEM (exceptfds);
+
+  return 0;
+}
+
+___HIDDEN int ___fdset_realloc
+   ___P((___processor_state ps,
+         int newsize),
+        (ps,
+         newsize)
+        ____processor_state ps;
+        int newsize;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+  int oldsize = ps->os.fdset.size;
+  int oldbytes;
+  int newbytes;
+
+  if (newsize <= oldsize) /* we don't shrink */
+    return 1;
+
+  oldbytes = ___CEILING_DIV (oldsize,8);
+  newbytes = ___CEILING_DIV (newsize,8);
+
+  readfds  = ___ALLOC_MEM (newbytes);
+  if (readfds == NULL)
+    goto error;
+
+  writefds = ___ALLOC_MEM (newbytes);
+  if (writefds == NULL)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (newbytes);
+  if (exceptfds == NULL)
+    goto error;
+#endif
+
+  memcpy (readfds, ps->os.fdset.readfds, oldbytes);
+  memset (___CAST(char*, readfds) + oldbytes, 0, newbytes - oldbytes);
+  memcpy (writefds, ps->os.fdset.writefds, oldbytes);
+  memset (___CAST(char*, writefds) + oldbytes, 0, newbytes - oldbytes);
+#ifdef USE_select
+  memcpy (exceptfds, ps->os.fdset.exceptfds, oldbytes);
+  memset (___CAST(char*, exceptfds) + oldbytes, 0, newbytes - oldbytes);
+#endif
+
+  ___FREE_MEM (ps->os.fdset.readfds);
+  ___FREE_MEM (ps->os.fdset.writefds);
+#ifdef USE_select
+  ___FREE_MEM (ps->os.fdset.exceptfds);
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = newsize;
+
+  return 1;
+
+ error:
+  if (readfds != NULL)
+    ___FREE_MEM (readfds);
+  if (writefds != NULL)
+    ___FREE_MEM (writefds);
+  if (exceptfds != NULL)
+    ___FREE_MEM (exceptfds);
+
+  return 0;
+}
+
+___HIDDEN int ___fdset_size
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ps->os.fdset.size;
+}
+
+___HIDDEN ___fdbits *___fdset_readfds
```

Comment:
These access functions are asymmetric, because there is no "setter". I suggest to remove these abstractions which depart from what is done in other parts of the runtime system.